stream_batch.go

```diff
@@ -0,0 +1,220 @@
+		confidenceWindowEnd := time.Now().Add(-sbr.config.Advanced.ConfidenceWindowSize)
+
+		if timeWindowEnd.After(confidenceWindowEnd) {
+			timeWindowEnd = confidenceWindowEnd
+		}
+
+		pollEnd := gocql.MinTimeUUID(timeWindowEnd)
+
+		var (
+			err      error
+			rowCount int
+		)
+
+		readUpTo := pollEnd
+
+		if compareTimeuuid(sbr.lastTimestamp, pollEnd) < 0 {
+			// Set the time interval from which we need to return data
+			var iter *changeRowIterator
+			iter, err = crq.queryRange(sbr.lastTimestamp, pollEnd)
+			if err != nil {
+				sbr.config.Logger.Printf("error while sending a query (will retry): %s", err)
+			} else {
+				var change Change
+				for {
+					streamCols, c := iter.Next()
+					if c == nil {
+						break
+					}
+
+					if c.GetOperation() == PreImage {
+						change.Preimage = append(change.Preimage, c)
+					} else if c.GetOperation() == PostImage {
+						change.Postimage = append(change.Postimage, c)
+					} else {
+						change.Delta = append(change.Delta, c)
+					}
+
+					if c.cdcCols.endOfBatch {
+						// Prevent changes older than the save point from being replayed again
+						if compareTimeuuid(oldProgress[string(change.StreamID)], streamCols.time) < 0 {
+							change.StreamID = streamCols.streamID
+							change.Time = streamCols.time
+							if err := consumer.Consume(change); err != nil {
+								// TODO: Does that make sense?
+								sbr.config.Logger.Printf("error while processing change (will quit): %s", err)
+								return err
+							}
+						} else {
+							sbr.config.Logger.Printf("skipping change due to it being too old (%v <= %v)",
+								streamCols.streamID, oldProgress[string(change.StreamID)].Time())
+						}
+
+						change.Preimage = nil
+						change.Delta = nil
+						change.Postimage = nil
+
+						// Update the last timestamp only after we processed whole batch
+						if compareTimeuuid(sbr.lastTimestamp, streamCols.time) < 0 {
+							sbr.lastTimestamp = streamCols.time
+						}
+					}
+
+					rowCount++
+				}
+
+				if err = iter.Close(); err != nil {
+					sbr.config.Logger.Printf("error while querying (will retry): %s", err)
+				}
+			}
+		} else {
+			// sbr.config.Logger.Printf("not polling")
+			readUpTo = sbr.lastTimestamp
+		}
+
+		sbr.lastTimestamp = readUpTo
+
+		if sbr.reachedEndOfTheGeneration(readUpTo) {
+			break outer
+		}
+
+		var delay time.Duration
+		if err != nil {
+			delay = sbr.config.Advanced.PostFailedQueryDelay
+		} else if rowCount > 0 {
+			delay = sbr.config.Advanced.PostNonEmptyQueryDelay
+		} else {
+			delay = sbr.config.Advanced.PostEmptyQueryDelay
+		}
+
+		delayUntil := time.Now().Add(delay)
+
+	delay:
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-time.After(delayUntil.Sub(time.Now())):
+				break delay
+			case <-sbr.interruptCh:
+				if sbr.reachedEndOfTheGeneration(readUpTo) {
+					break outer
+				}
+			}
+		}
+	}
+
+	for _, stream := range sbr.streams {
```

Comment:
shouldn't we be saving progress periodically?
If we only save at the end of generation or at graceful close (afaik that's what this currently does), it's easy to lose a lot of progress if there was a single generation operating for a very long time

Reply:
Yes, it definitely should. I thought we could allow `ChangeConsumer` to decide when progress should be saved. For example, it would receive a `ProgressReporter` object which would have a method `MarkProgress`. The drawback of this method is that the consumer won't be able to save progress if there are no changes for a long time. WDYT?

Reply:
we could timeout and call the consumer with an empty sequence of changes

Reply:
So, an empty `Change` object? What would be its properties?

Or should we maybe change the signature of `(ChangeConsumer).Consume` to take a slice of `Change`s, not a single change? Then the empty sequence of changes would mean an empty slice.

Or, maybe `ChangeConsumer` could receive a reference to a `ProgressReporter` when it's created, and the user could implement a sidecar goroutine which would take care of saving the progress when nothing happens?

Reply:
The last option seems to be the most flexible, but in that case I'd require `ProgressReporter.report` (or whatever its called) to be thread-safe; also, would require including that in the example replicator

Reply:
I modified the replicator to use the `ProgressReporter`, and it asynchronously saves progress via a helper struct (`PeriodicProgressReporter`).